File: FutureVoice/Views/FeedbackSheet.swift
import SwiftUI
import Supabase

/// The ONE feedback modal, asked ONCE: two scores and a sentence, after a call
/// that actually happened, from someone who came back to have it.
///
/// **It used to fire on the first call and the first Watch, and that was the
/// wrong moment twice over** (changed 2026-09-14). It interrupted the single
/// experience the whole product is selling, at the exact second the learner
/// had just done the thing — and it asked for a verdict from someone who had
/// nothing to compare it to. A first impression is not an opinion. What
/// replaced both is one ask at the first moment a learner has actually
/// decided something: they had a call, they left, they CAME BACK, and they
/// have just finished another real one (`FeedbackPrompt.shouldShowReturningTalk`).
///
/// **Named for the beta until 2026-08-18, kept for the public build.** Asking
/// someone what felt off right after their first call is worth as much at
/// launch as it was in testing — what had to go was the word "beta", which
/// told a paying user they'd wandered into someone else's test. Two things
/// deliberately did NOT get renamed with it:
///
/// - **The `beta_reviews` table.** Nobody sees a table name, and renaming it
///   means a migration that has to land in production BEFORE the build does —
///   ship them out of order and every piece of feedback is silently dropped
///   on insert. Not a trade worth making for a cosmetic rename.
/// - **The `futurevoice.betaFeedback.*` UserDefaults keys.** They're the
///   record of who has already been asked. New keys would read as "never
///   asked" and re-prompt every existing user on their next call.
///
/// **Stars, on ONE of the three asks** (2026-08-17 removed them, 2026-09-14
/// brought them back for `returningTalk` only). The original objection was
/// about the MOMENT, not the control: a 1–5 row right after someone's very
/// first call reads as an App Store review prompt, and a score given before
/// there is anything to score tells us nothing — a 2 with no sentence is
/// unusable and a 5 hides what felt off. That still holds for the first call
/// and the first Watch, which stay words-only. By the second call on a return
/// visit it doesn't: they have used the thing, and two numbers across many
/// people is the one signal a pile of sentences cannot give. They are asked
/// separately — the APP and the CALLS — because someone can love the calls
/// and find everything around them confusing, which is precisely the feedback
/// worth having, and a single score blurs it.
///
/// Presentation is once-per-milestone, tracked in UserDefaults via
/// `FeedbackPrompt.shouldShow` / `markShown` — callers decide the moment,
/// this file decides the look.
struct FeedbackSheet: View {
    /// The milestone that triggered the ask. Raw values are the `context`
    /// column in `beta_reviews` — never rename one, it would split the
    /// history of a question that didn't change.
    enum Context: String, Identifiable {
        /// A learner who came BACK and had another real call. The only ask
        /// there is — see the type doc for the two it replaced, whose rows
        /// (`first_talk`, `first_watch`) stay in the table under their own
        /// names.
        case returningTalk = "returning_talk"

        var id: String { rawValue }

        var title: String { explain("How is it going so far?") }

        var subtitle: String {
            explain("You've been back for another call. Two quick questions, and anything you want to say.")
        }
    }

    let context: Context

    @Environment(\.dismiss) private var dismiss
    @State private var feedbackText = ""
    @State private var appRating = 0
    @State private var callRating = 0
    @State private var sending = false
    @State private var sendError: String?
    @State private var sent = false

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    VStack(alignment: .leading, spacing: 8) {
                        Text(context.title)
                            .font(.title2.bold())
                            .fixedSize(horizontal: false, vertical: true)
                        Text(context.subtitle)
                            .font(.callout)
                            .foregroundStyle(.secondary)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.vertical, 4)
                    .listRowBackground(Color.clear)
                }

                Section {
                    StarRow(title: explain("The app overall"), rating: $appRating)
                    StarRow(title: explain("Talking to your future self"), rating: $callRating)
                }

                Section(explain("Anything you want to say")) {
                    TextEditor(text: $feedbackText)
                        .frame(minHeight: 110)
                }

                if let sendError {
                    Section {
                        Label(sendError, systemImage: "exclamationmark.triangle")
                            .foregroundStyle(.red)
                    }
                }
            }
            .navigationTitle("")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Not now") { dismiss() }
                }
                ToolbarItem(placement: .topBarTrailing) {
                    if sending {
                        ProgressView()
                    } else {
                        Button("Send") { Task { await submit() } }
                            .disabled(!canSend)
                    }
                }
            }
            .interactiveDismissDisabled(sending)
            .alert("Thank you", isPresented: $sent) {
                Button("Done") { dismiss() }
            } message: {
                Text(explain("Your feedback goes straight to the person building this."))
            }
        }
        .presentationDetents([.medium, .large])
    }

    /// A score on its own is a complete answer, and so is a sentence on its
    /// own. Demanding both is how a form gets neither.
    private var canSend: Bool {
        if !feedbackText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return true }
        return appRating > 0 || callRating > 0
    }

    private func submit() async {
        guard let session = try? await SupabaseProvider.shared.auth.session else {
            sendError = explain("You need to be signed in.")
            return
        }
        sending = true
        sendError = nil

        // `rating` is the APP score and `call_rating` the CALLS score; both
        // are omitted when unset (a nil Optional encodes to no key at all, so
        // the column keeps its null). `first_talk` / `first_watch` rows still
        // carry neither, exactly as before.
        struct Row: Encodable {
            let user_id: String
            let context: String
            let body: String
            let rating: Int?
            let call_rating: Int?
            let app_version: String?
            let app_build: String?
        }
        let info = Bundle.main.infoDictionary
        do {
            try await SupabaseProvider.shared
                .from("beta_reviews")     // see the type doc — name kept on purpose
                .insert(Row(
                    user_id: session.user.id.uuidString,
                    context: context.rawValue,
                    body: feedbackText.trimmingCharacters(in: .whitespacesAndNewlines),
                    rating: appRating > 0 ? appRating : nil,
                    call_rating: callRating > 0 ? callRating : nil,
                    app_version: info?["CFBundleShortVersionString"] as? String,
                    app_build: info?["CFBundleVersion"] as? String
                ))
                .execute()
            sent = true
        } catch {
            sendError = explain("Couldn't send — please try again. (\(error.localizedDescription))")
        }
        sending = false
    }
}

/// One line of the rating section: a label and five taps.
///
/// Built from `Button` + SF Symbols rather than a control of its own — the
/// house rule is iOS-native elements only, and a star row has no system
/// control. Tapping the star you already chose clears it: a score given by
/// accident must be takeable back, and both scores are optional.
private struct StarRow: View {
    let title: String
    @Binding var rating: Int

    var body: some View {
        HStack {
            Text(title)
            Spacer(minLength: 12)
            HStack(spacing: 2) {
                ForEach(1...5, id: \.self) { value in
                    Button {
                        rating = (rating == value) ? 0 : value
                    } label: {
                        Image(systemName: value <= rating ? "star.fill" : "star")
                            .font(.title3)
                            .foregroundStyle(value <= rating ? AnyShapeStyle(.tint)
                                                             : AnyShapeStyle(.secondary))
                            .frame(width: 30, height: 30)
                            .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("\(value)")
                }
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityValue(rating == 0 ? explain("Not rated")
                                        : explain("\(rating) out of 5"))
    }
}

/// Once-per-milestone bookkeeping. Callers check `shouldShow` at the moment
/// the milestone completes, then `markShown` before presenting.
enum FeedbackPrompt {
    /// Key prefix is the beta-era one on purpose — it's the record of who has
    /// already been asked, and a new prefix would re-prompt everyone.
    private static func key(_ c: FeedbackSheet.Context) -> String {
        "futurevoice.betaFeedback.\(c.rawValue)"
    }

    static func shouldShow(_ c: FeedbackSheet.Context) -> Bool {
        !UserDefaults.standard.bool(forKey: key(c))
    }

    static func markShown(_ c: FeedbackSheet.Context) {
        UserDefaults.standard.set(true, forKey: key(c))
    }

    /// The one moment worth asking a learner what they think: they had a call,
    /// they came BACK, and they have just finished another real one.
    ///
    /// Three conditions, and each rules out a kind of answer that would be
    /// noise:
    ///
    /// - **A minute of call.** Under that nothing happened — an accidental
    ///   tap, a call abandoned on the greeting — and an opinion about it is an
    ///   opinion about nothing. The clock is the call's own elapsed time, the
    ///   figure the screen was showing them.
    /// - **At least their second finished call**, counted across languages
    ///   (the current one is already saved by the time this is asked, so two
    ///   means this one plus an earlier one).
    /// - **They came back.** The first call has to be on an EARLIER DAY than
    ///   today. Two calls in one sitting is still a first impression; coming
    ///   back tomorrow is the first evidence the thing is worth returning to,
    ///   and it is the only point where "how is it going" is a real question.
    ///
    /// Asked once, ever — the same UserDefaults record as the other two.
    static func shouldShowReturningTalk(callSeconds: TimeInterval) -> Bool {
        guard callSeconds >= 60, shouldShow(.returningTalk) else { return false }
        let ended = SessionStore.shared.loadAcrossLanguages()
            .filter { $0.endedAt != nil }
        guard ended.count >= 2,
              let first = ended.min(by: { $0.startedAt < $1.startedAt })
        else { return false }
        return !Calendar.current.isDateInToday(first.startedAt)
    }
}

File: android/app/src/main/java/com/roro/futurevoice/ui/FeedbackSheet.kt
package com.roro.futurevoice.ui

import android.content.Context
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.roro.futurevoice.R
import com.roro.futurevoice.core.Config
import com.roro.futurevoice.data.AuthRepository
import com.roro.futurevoice.net.Edge
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * The ONE feedback modal — shown at milestone moments (first conversation
 * ended, first Watch listen-through), always the same shape: the question,
 * a text box, send. No star row: a score right after the first call reads
 * like a store-review prompt and tells us nothing actionable; the sentence
 * is what we need. Raw values are the `context` column in `beta_reviews` —
 * never rename one, it would split the history of a question that didn't
 * change.
 */
enum class FeedbackContext(val raw: String, val title: Int, val subtitle: Int) {
    FIRST_TALK("first_talk", R.string.how_was_your_first_conversation,
        R.string.you_just_talked_with_your_future_voice_what_felt_right_and_w_7a6d31),
    FIRST_WATCH("first_watch", R.string.how_was_your_first_watch,
        R.string.you_just_heard_yourself_handle_a_real_situation_did_it_sound_ed28b5),
}

/** Once per milestone. Key prefix is the beta-era one on purpose — it is the
 *  record of who has already been asked. */
object FeedbackPrompt {
    private fun key(c: FeedbackContext) = "futurevoice.betaFeedback.${c.raw}"
    fun shouldShow(context: Context, c: FeedbackContext) =
        !context.getSharedPreferences("futurevoice", 0).getBoolean(key(c), false)
    fun markShown(context: Context, c: FeedbackContext) =
        context.getSharedPreferences("futurevoice", 0).edit().putBoolean(key(c), true).apply()
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FeedbackSheet(context: FeedbackContext, onDismiss: () -> Unit) {
    val app = LocalContext.current
    val scope = rememberCoroutineScope()
    var text by remember { mutableStateOf("") }
    var sending by remember { mutableStateOf(false) }
    var sendError by remember { mutableStateOf<String?>(null) }
    var sent by remember { mutableStateOf(false) }

    if (sent) {
        AlertDialog(
            onDismissRequest = onDismiss,
            title = { Text(stringResource(R.string.thank_you)) },
            text = { Text(stringResource(R.string.your_feedback_goes_straight_to_the_person_building_this)) },
            confirmButton = { TextButton(onClick = onDismiss) { Text(stringResource(R.string.done)) } },
        )
        return
    }
    ModalBottomSheet(onDismissRequest = { if (!sending) onDismiss() }) {
        Column(Modifier.fillMaxWidth().navigationBarsPadding().padding(horizontal = 20.dp).padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)) {
            Text(stringResource(context.title), style = MaterialTheme.typography.titleLarge, fontWeight = FontWeight.Bold)
            Text(stringResource(context.subtitle), style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
            OutlinedTextField(value = text, onValueChange = { text = it },
                label = { Text(stringResource(R.string.your_feedback)) },
                modifier = Modifier.fillMaxWidth().heightIn(min = 110.dp))
            sendError?.let { Text(it, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.error) }
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                TextButton(onClick = onDismiss, enabled = !sending, modifier = Modifier.weight(1f)) { Text(stringResource(R.string.not_now)) }
                Button(onClick = {
                    sending = true; sendError = null
                    scope.launch {
                        runCatching { send(context, text.trim()) }
                            .onSuccess { sent = true }
                            .onFailure { sendError = app.getString(R.string.feedback_couldnt_send, it.message ?: "") }
                        sending = false
                    }
                }, enabled = !sending && text.isNotBlank(), modifier = Modifier.weight(1f)) {
                    if (sending) CircularProgressIndicator(Modifier.size(18.dp), strokeWidth = 2.dp)
                    else Text(stringResource(R.string.send))
                }
            }
            Spacer(Modifier.size(4.dp))
        }
    }
}

/** No `rating` — the column is nullable and the sheet doesn't ask for a score. */
private suspend fun send(context: FeedbackContext, body: String) = withContext(Dispatchers.IO) {
    val auth = AuthRepository()
    val uid = auth.userId ?: throw IllegalStateException("signed out")
    val row = buildJsonObject { put("user_id", uid); put("context", context.raw); put("body", body) }
    val req = Request.Builder().url(Config.supabaseUrl.trimEnd('/') + "/rest/v1/beta_reviews")
        .header("Authorization", "Bearer ${auth.accessToken()}").header("apikey", Config.supabaseAnonKey)
        .header("Prefer", "return=minimal")
        .post(Edge.json.encodeToString(JsonObject.serializer(), row).toRequestBody("application/json".toMediaType()))
        .build()
    Edge.client.newCall(req).execute().use { r -> if (r.code !in 200..299) throw IllegalStateException("HTTP ${r.code}") }
}
